src/queue.c

```diff
@@ -0,0 +1,46 @@
+#include "queue.h"
+
+#ifdef __arm__
+#define MG_MEMORY_BARRIER() asm("dsb; isb;")
+#else
+#define MG_MEMORY_BARRIER() (void) 0
+#endif
+
+// Align given value up
+static unsigned long mg_align(unsigned long value, unsigned long align) {
+  return (value + align - 1) / align * align;
+}
+
+unsigned long mg_qspace(struct mg_queue *q, unsigned long *ofs) {
+  *ofs = q->tail >= q->head ? sizeof(unsigned long)
+                            : q->head + sizeof(unsigned long);
+  return *ofs > q->len ? 0 : q->len - *ofs;
+}
+
+void mg_qproduce(struct mg_queue *q, unsigned long head, unsigned long len) {
+  unsigned long newhead = head + mg_align(len, sizeof(unsigned long));
+  if (newhead <= q->len) {
+    *(unsigned long *) (q->buf + head - sizeof(unsigned long)) = len;
+    if (head == sizeof(unsigned long)) {
+      q->head = 0;          // Reset head first, to prevent consumer from read
+      MG_MEMORY_BARRIER();  // Prevent instruction reordering
+      q->tail = 0;          // Now reset the tail
+    }
+    q->head = newhead;
+  }
+}
+
+unsigned long mg_qnext(struct mg_queue *q, unsigned long *ofs) {
+  *ofs = q->head > q->tail ? q->tail + sizeof(unsigned long) : 0;
+  return *ofs ? *(unsigned long *) &q->buf[*ofs - sizeof(unsigned long)]
+              : MG_QUEUE_EMPTY;
+}
+
+void mg_qconsume(struct mg_queue *q) {
+  unsigned long tail, len = mg_qnext(q, &tail);
+  if (len > 0) {
+    tail += mg_align(len, sizeof(unsigned long));
+    if (tail >= q->head) tail = q->head;
+    q->tail = tail;
+  }
+}
```

Comment:
This is a bit convoluted... (for me) and it's late, so please excuse if I write nonsense

You ask for room. Then you pass that as a parameter. Once there is room, that won't change, it can only get better when the consumer does his thing; so, once there is room, return the amount, and q-> head is still the place to put that (if it fits); unless I'm not seeing something, there is no need to pass parameter head to qproduce()

q->tail should never be greater than q->head, at least equal, and that happens either when the queue is empty and when the queue is full, that is why some people leave one byte "reserved", to prevent the tail = head condition on queue full. With this as it is, you may purge a full queue if no msg has been consumed when it fills, as you see that as an empty queue

You push messages and the head advances, you pop messages and the tail advances, when head is at the end of the queue, even though the queue may be empty, you return only the room between head and end. Many frames will be dropped until one fits that size. You should reset head when there is no room for a whole frame AND tail is above that (head > len - MAX_PKT_SIZE -  sizeof(UL)) && (tail >= MAX_PKT_SIZE + sizeof(UL))

Reply:
I am not sure you're correct. `you return only the room between head and end` - that's wrong.

When tail == head, it returns the full queue length:
```c
unsigned long mg_qspace(struct mg_queue *q, unsigned long *ofs) {
  *ofs = q->tail >= q->head ? sizeof(unsigned long)
                            : q->head + sizeof(unsigned long);
  return *ofs > q->len ? 0 : q->len - *ofs;
}
```

Anyway, what's your suggested API?